lib/schema/src/openctx.schema.ts

```diff
@@ -1,18 +1,66 @@
 /**
  * Metadata about code
  */
-export type Schema = Item | Annotation
+export type Schema = CandidateItem | Item | Annotation
 /**
  * A hint about how to best present an annotation to the human in the client's user interface.
  *
  * - `prefer-link-over-detail`: Prefer to show the item as a link over showing the detail text, if the client does not cleanly support doing both.
  */
 export type PresentationHint = 'prefer-link-over-detail'
+/**
+ * Additional information which can be used to idenfity the resource by the provider.
+ */
+export type ItemMeta = { [key: string]: any }
 
+export type ItemKindID = string
+
+/**
+ * A kind of context item returnable by the provider.
+ */
+export interface ItemKind<K extends ItemKindID = ItemKindID> {
+    /**
+     * The identifier for the item kind.
+     */
+    id: K
+    /**
+     * The human-readable name of the item kind.
+     */
+    name: string
+    /**
+     * A human-readable description of the item kind.
+     */
+    description?: string
```

Comment:
What if 2 providers both return items of the same kind? I wonder if we should treat providers as a concept in the UI. Should we show the list of registered providers in the @-mention list, or should we show the kinds?

I think we should start with the former. That requires less "figuring out how to shove the provider author's semantic model of the world into consistency with that of other provider authors and us". They name their provider and then they're done. With item kinds, how do you deal with ItemKind.id conflicts? Should ItemKind.name be `Google Doc` or `Google Drive document` or `RFC`? It just requires an extra step of ascribing meaning that we don't know enough about yet as API designers.

Reply:
Listing the providers from the client is a TODO and we definitely want to do that. But we will need a 2-level grouping. A single provider like Github, can support different item kinds like PRs & Issues, same with Jira or other 3rd party APIs. 

One way to render the list is:
```javascript
client.capabilities().flatMap(provider => provider.itemKinds.map(kind => 
  <div key={`${provider.uri}-${kind.id}`}>
    {kind.title} <sub>(by ${provider.title})</sub>
  </div>
))
```
```
-----------------------------
Github Issues (by Github)
-----------------------------
Github PRs (by Github)
-----------------------------
```

Second Option 
```javascript
client.capabilities().map(provider => 
  <div key={provider.uri} onClick={() => setSelected(provider)}>
    {provider.title}
  </div>
)

selected && selected.itemKinds.map(kind => <div key={kind.id}>{kind.title}</div>)
```
```
-----------------------------|-----------------------------
Github                       | Issues
-----------------------------|-----------------------------
                             | PRs
                             |-----------------------------
```

Reply:
@sqs the main motivation behind this is I think it is reasonable for a provider to provide multiple types of context that a user may want to filter on. eg github provider does issues and PRs. A sourcegraph search provider could provide _lots_ of types (commits, packages, type:file, global (via dotcom), code intel?). I imagine a large customer would have lots of internal systems they may want to integrate with, and would probably deploy as a single provider?

Reply:
I see. And if we want to have a multiplexer here then we need one protocol counterparty to be able to represent stuff from multiple providers. Ie the provider is itself not a meaningful concept, there is no provider-specific implicit state.

This justifies a `kind` for me. So items would have `kind?: string` and capabilities would have `itemKinds: {id: string, title?: string}[]`? That makes sense to me.

I could see an argument for capabilities `itemKinds` being just `string[]` and being presentational, but the item kind `title` may need to be dynamic or we may want to have a separate icon/description field in the future. Neither of those are pressing needs, but they seem likely enough that we should make it an object with `{id: string, title?: string}` for now. It's annoying in a lot of languages to JSON-decode a `string | { id: string, ...}` value.

Reply:
Thanks!

I think we are clear here. So resolving this thread.